File: code/problems/n_queens/s2.rb
def n_queens(n)
  board = Array.new(n) { Array.new(n, 0) }
  diagonals = [[-1, -1], [1, -1], [1, 1], [-1, 1]]

  can_place_queen = lambda do |row, column|
    (0 .. n-1).each do |i|
      return false if board[row][i] == 1
      return false if board[i][column] == 1
    end
    diagonals.each do |diagonal|
      i, j = row + diagonal[0], column + diagonal[1]
      while i >= 0 and i < n and j >= 0 and j < n
        return false if board[i][j] == 1
        i, j = i + diagonal[0], j + diagonal[1]
      end
    end
    true
  end

  try_row = lambda do |row|
    return true if row == n
    (0 .. n-1).each do |column|
      next unless can_place_queen.call(row, column)
      board[row][column] = 1
      if try_row.call(row + 1)
        return board
      else
        board[row][column] = 0
      end
    end
    false
  end

  try_row.call(0)
end

File: code/problems/n_queens/s2.js
export function nQueens(n) {
  const board = Array(n).fill().map(() => Array(n).fill(0))
  const directions = [
    [0,1], [1,1], [1,0], [1,-1],
    [0,-1], [-1,-1], [-1,0], [-1,1]
  ]
  const canPlaceQueen = (i, j) => {
    for (let k = 0; k < 8; k++) {
      let direction = directions[k]
      let p = [i + direction[0], j + direction[1]]
      while (p[0] >= 0 && p[0] < n && p[1] >= 0 && p[1] < n) {
        if (board[p[0]][p[1]] === 1) {
          return false
        }
        p[0] += direction[0]
        p[1] += direction[1]
      }
    }
    return true
  }
  const tryRow = (i) => {
    if (i === n) {
      return true
    }
    for (let j = 0; j < n; j++) {
      if (!canPlaceQueen(i, j)) {
        continue
      }
      board[i][j] = 1
      if (tryRow(i + 1)) {
        return true
      } else {
        board[i][j] = 0
      }
    }
  }
  return tryRow(0) ? board : false
}
